### PR TITLE
Introduce `stub` mode for `-unavailable-decl-optimization`

### PR DESCRIPTION
**Pull request by tshortli**
Implement a new `-unavailable-decl-optimization` mode that replaces the bodies of unavailable functions with calls to a standard library function that invokes `fatalError`. Unlike `-unavailable-decl-optimization=complete`, this compilation mode is designed to improve the code size of unavailable code while also preserving ABI compatibility. This is important since there are existing compiled binaries that link the symbols for unavailable functions even though those symbols should be provably unreachable.

This optimization is implemented in SILGen by inserting a call to a new standard library function `_diagnoseUnavailableCodeReached()` in the prologue of each unavailable function. The existing dead code elimination optimization pass then determines that the remainder of the function body is unreachable and removes all other code, leaving the body of the function as a single call to a no-return function.

There are several follow up changes that can be made to maximize the effectiveness of this optimization:
- Some functions that could benefit from this optimization are generated during IRGen, like value witnesses functions. The same technique will need to be reproduced for IRGen to reduce code size further.
- The generated machine code for the stubbed functions is not optimal. It currently includes stack pointer management before calling `_diagnoseUnavailableCodeReached()` and sometimes also includes a `brk` instruction after the call. Some work to reduce the machine code for these stubs down to the minimal necessary instructions (probably just a single branch?) may be warranted.

Resolves rdar://107388493

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test Windows

**Comment by tshortli** [comment]
@swift-ci please test

File: include/swift/SIL/SILInstruction.h [hunk 1]
```diff
@@ -2777,6 +2777,14 @@ class ApplyInstBase<Impl, Base, false> : public Base {
     return calleeFn->hasSemanticsAttr(SEMANTICS_PROGRAMTERMINATION_POINT);
   }
 
+  /// Returns true if the callee function is annotated with
+  /// @_semantics("unavailable_code_reached")
+  bool isCalleeUnavailableCodeReached() const {
+    auto calleeFn = getCalleeFunction();
+    if (!calleeFn) return false;
```
Comment (nkcsgexi) [inline include/swift/SIL/SILInstruction.h:2784]: nitpick: usually we split `return` to a separate line for better debuggability.

**Comment by tshortli** [comment]
@swift-ci please test Linux

**Pull request merged**